WAF: Add Waf_Manager Class

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

The objective of this PR is to reduce the amount of logic inside the `Waf_Runner` class that isn't solely applicable to the actual execution of the web application firewall during a site request.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Extracts some higher-level logic from the `Waf_Runner` class into a new `Waf_Manager` class.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

peb6dq-mP-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Proof read changes.
* Double check tests are still passing.
* Smoke test the firewall feature in Jetpack and Jetpack Protect - this PR should introduce no changes in functionality.

dkmyta commented on projects/packages/waf/src/class-waf-manager.php:
```diff
@@ -0,0 +1,205 @@
+	 */
+	public static function activate() {
+		Waf_Constants::define_mode();
+		if ( ! self::is_allowed_mode( JETPACK_WAF_MODE ) ) {
```
Looks like `is_allowed_mode()` is still currently a `Waf_Runner` method.

After updating that, however, I was hit with a separate fatal:
```
PHP Fatal error:  Uncaught Exception: Failed writing allow list file to: /var/www/html/wp-content/jetpack-waf/rules/allow-ip.php in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-rules-manager.php:312
```
This may have just been a product of making changes on the fly to the WAF package without installing after the updates as I couldn't replicate it consistently. In summary, this happened on a fresh site (no existing WAF options) after installing/activating Jetpack, connecting my user, activating the Firewall, and fixing the initial fatal.

dkmyta commented on projects/packages/waf/src/class-waf-manager.php:
```diff
@@ -0,0 +1,205 @@
+	 * @throws \Exception If file writing fails.
+	 */
+	public static function deactivate() {
+		delete_option( self::MODE_OPTION_NAME );
```
Another `Waf_Runner` constant.

dkmyta commented on projects/packages/waf/src/class-waf-manager.php:
```diff
@@ -0,0 +1,205 @@
+			add_option( Waf_Rules_Manager::VERSION_OPTION_NAME, Waf_Rules_Manager::RULES_VERSION );
+		}
+
+		add_option( self::SHARE_DATA_OPTION_NAME, true );
```
This also appears to be a `Waf_Runner` constant.

dkmyta commented on projects/plugins/protect/src/class-jetpack-protect.php:
```diff
@@ -202,7 +202,7 @@ public function initial_state() {
 				'isSupported' => Waf_Runner::is_supported_environment(),
 				'isSeen'      => self::get_waf_seen_status(),
-				'isEnabled'   => Waf_Runner::is_enabled(),
+				'isEnabled'   => Waf_Manager::is_enabled(),
```
We'll need to add a top-level `use` statement for this new class:
```
use Automattic\Jetpack\Waf\Waf_Manager;
```

dkmyta reviewed (dismissed):
With the exception of the random fatal I couldn't replicate consistently regarding the failure to write to the allow list (that may not necessarily be directly related to these changes), with all suggested updates applied, everything looks great and appears to continue to work nicely.

There is also currently one WAF-related post-connection test failing that will require some attention.

nateweller committed:
Add Waf_Manager class

nateweller committed:
Changelogs

nateweller commented:
@dkmyta Thanks for the initial review! I've rebased on `trunk` and made some updates.

The only notable differences from your last review should be:
1. Moved all option names into `Waf_Manager` - no need for them to be in `Waf_Runner`.
2. Moved the new `automatic_rules_available()` method from `Waf_Runner` to `Waf_Rules_Manager`.

Otherwise everything should be the same :+1:

nateweller committed:
Update README (#28107)

nateweller commented:
^ Also merged in a README update, which Miguel approved here: #28107

dkmyta reviewed (changes requested):
I think perhaps rebasing has reintroduced the initial issues with the `Waf_Manager` using `self::` for a `Waf_Runner` method - in particular `is_allowed_mode()`. The other uses for the constants are no longer an issue.

nateweller commented:
Going to close this and come back to it after we have finished updating the WAF with improved error handling and test coverage.
